### PR TITLE
add cuda compiler and rerender

### PR DESCRIPTION
**Pull request by izahn**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by izahn** [comment]
This isn't actually ready for review!

I'm working on a cuda build, but it goes off the rails right from the start. The build logs in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=323356&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642 show that just adding `- {{ compiler('cuda') }}  # [cuda_compiler_version != "None"]` to the build requirements is enough to break it. The error message is 

> ERROR: /home/conda/.cache/bazel/_bazel_conda/4396700a6d395f5dd5422cba677ec32d/external/com_google_absl/absl/base/BUILD.bazel:54:1: The include path '/usr/local/cuda/include' references a path outside of the execution root.

I've tried a bunch of things over in https://github.com/izahn/tensorflow-feedstock/tree/cuda2 but I haven't been able to figure it out.

In case it helps at all, I do have a working cuda build over in https://github.com/izahn/tensorflow-feedstock/tree/anaconda based on https://github.com/AnacondaRecipes/tensorflow_recipes but it is so radically different from the conda-forge version that I hesitate to submit anything based on it here.

**Comment by njzjz** [comment]
You need to manually set the cuda path. See https://github.com/AnacondaRecipes/tensorflow_recipes/blob/6fa2f061cc688265f98af2f5bceccc92cee6d50c/tensorflow-base-gpu/build.sh#L55

**Comment by izahn** [comment]
I set the cuda path in https://github.com/conda-forge/tensorflow-feedstock/pull/118/commits/1e45ce5e13c1f9839944ea3dabc923b8e15b8b56 but the error is the same. Any idea what I'm doing wrong?

**Comment by izahn** [comment]
Also, even https://github.com/izahn/tensorflow-feedstock/tree/cuda2 fails the same way, despite setting `CUDA_TOOLKIT_PATH`, `TF_CUDA_PATHS`, `TF_CUDA_VERSION` and `TF_CUDNN_VERSION`, adapted from https://github.com/AnacondaRecipes/tensorflow_recipes. I'm sure it will seem obvious once we find it, but I've been trying for a while with no luck.

**Comment by njzjz** [comment]
let me take a look... I can still see the following in the log
```
2021-05-25T09:53:33.6702369Z + export 'CC_OPT_FLAGS=-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1621935457837/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1621935457837/work=/usr/local/src/conda/tensorflow-split-2.4.1 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1621935457837/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac=/usr/local/src/conda-prefix **-I/usr/local/cuda/include**'
```

**Comment by njzjz** [comment]
Well I find the source of this flag. It comes from nvcc https://github.com/conda-forge/nvcc-feedstock/blob/25f322292e8ac423b25af323199caeb2f307460c/recipe/install_nvcc.sh#L72-L75.
So do we need to keep this flag?

**Comment by xhochy** [comment]
You will need to add it as a `cxx_builtin_include_directories` as part of the Bazel toolchain: https://github.com/conda-forge/tensorflow-feedstock/blob/f18fb006b85678a5180336e5f99f68a9ba20d5ea/recipe/custom_toolchain/cc_toolchain_config.bzl#L296-L310

**Comment by izahn** [comment]
> You will need to add it as a `cxx_builtin_include_directories` as part of the Bazel toolchain:

Yep, that's the ticket thank you!

File: recipe/custom_toolchain/cc_toolchain_config.bzl [hunk 1]
```diff
@@ -307,6 +307,8 @@ def _impl(ctx):
 	    "${BUILD_PREFIX}/lib/gcc/${HOST}/${COMPILER_VERSION}",
 	    "${BUILD_PREFIX}/${HOST}/include/c++/${COMPILER_VERSION}",
             "${PREFIX}/include",
+            "/usr/local/cuda/include",
+            "/usr/local/cuda-${cuda_compiler_version}/include",
```
Comment (xhochy) [inline recipe/custom_toolchain/cc_toolchain_config.bzl:311]: Note that we pass in the variables somewhere explicitly. They are not automatically loaded from the environment.

File: recipe/build.sh [hunk 2]
```diff
@@ -110,6 +110,8 @@ export TF_CONFIGURE_IOS=0
 sed -i -e "/PROTOBUF_INCLUDE_PATH/c\ " .bazelrc
 sed -i -e "/PREFIX/c\ " .bazelrc
 
+export TF_CUDA_PATHS="${PREFIX},/usr/local/cuda-${cuda_compiler_version},/usr"
```
[the remark this answers is not in the document]
Comment (izahn) [reply recipe/build.sh]: Now we get `/home/conda/feedstock_root/build_artifacts/tensorflow-split_1621962893444/work/conda_build.sh: line 116: CUDA_HOME: unbound variable`, I'm going to revert this.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -35,6 +35,9 @@ requirements:
     - nasm
     - sed
   host:
+    # GPU requirements
+    - cudnn                           # [cuda_compiler_version != "None"]
+    - nccl                            # [cuda_compiler_version != "None"]
```
Comment (njzjz) [inline recipe/meta.yaml:40]: I think you forgot to add [cudatoolkit-dev](cudatoolkit-dev-feedstock)?

Comment (izahn) [reply recipe/meta.yaml:40]: I'm using https://github.com/conda-forge/pytorch-cpu-feedstock/blob/master/recipe/meta.yaml and https://github.com/AnacondaRecipes/tensorflow_recipes/blob/master/tensorflow-base-gpu/meta.yaml as templates, neither of which has cudatoolkit-dev. I also don't see any mention of it at https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=cuda#cuda-builds

Comment (jakirkham) [reply recipe/meta.yaml:40]: Yeah `cudatoolkit-dev` isn't used for this `{{ compiler("cuda") }}` is (per the docs)

Comment (njzjz) [reply recipe/meta.yaml:40]: You're right. I find the docs: https://conda-forge.org/docs/user/faq.html?highlight=cuda#faq-cuda-compiler-header

**Comment by jakirkham** [comment]
Maybe increase the verbosity of `./configure`? It's doing something. We just need a better idea of what :)

**Comment by xhochy** [comment]
> I've been stuck for a while on the ./configure step. There's no error, it just stops and hangs forever at

That is an indicator that it is awaiting an active input on the command line. I suggest to rather debug this locally and not using CI.

**Comment by izahn** [comment]
Well I've come full circle. Figured out the `configure` hang was because `GCC_HOST_COMPILER_PATH` needs to be set, but now we're back to 

```
INFO: Analyzed 3 targets (402 packages loaded, 26378 targets configured).
INFO: Found 3 targets...
[0 / 209] [Prepa] BazelWorkspaceStatusAction stable-status.txt
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622019308407/work/tensorflow/stream_executor/BUILD:649:1: The include path '/usr/local/cuda/include' references a path outside of the execution root.
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1622019308407/work/tensorflow/python/BUILD:6068:1 The include path '/usr/local/cuda/include' references a path outside of the execution root.
INFO: Elapsed time: 42.268s, Critical Path: 1.18s
```

**Comment by izahn** [comment]
This is pretty much a mess at this point, closing and will open a clean PR in a minute.

**Pull request closed without merging**